[Stdlib] Make Sequence.SubSequence conform to Sequence.

DougGregor opened the pull request:
Sitting on top of https://github.com/apple/swift/pull/11751, this PR places the "obvious" recursive requirements on `Sequence.SubSequence`:

* It is a `Sequence`
* It's `Element` type matches that of the `Sequence`
* Slicing a `SubSequence` produces the same `SubSequence`.

Fixes [SR-318](https://bugs.swift.org/browse/SR-318) / rdar://problem/31418206.

DougGregor committed:
[GSB] Infer equivalence of same-named nested types of the same potential archetype.

When we have two nested types of a given potential archetype that have
the same name, introduce a (quietly) inferred constraint. This is
a future-proofing step for canonicalization, for a possible future
where we no longer require all of the nested types of a given name
to be equivalent, e.g., because we have a proper disambiguation
mechanism.

DougGregor committed:
[GSB] Remove a spurious variable update. NFC

DougGregor committed:
[GSB] Eliminate self-derived nested-type-name-match constraints..

Nested-type-name-match constraints are odd because they are
effectively artifacts of the GenericSignatureBuilder's algorithm,
i.e., they occur when we have two PotentialArchetypes representing the
same type, and each of those PA's has a nested type based on the same
associated type. Because of this, nested-type-name-match constraints
have no useful requirement sources, so the normal means of detecting
self-derived constraints doesn't suffice, and we end up with
self-derived nested-type-name-match constraints eliminating the
constraints they depend on, causing spurious "redundant same-type
constraint" diagnostics and minimized generic signatures that drop
important requirements.

Handle nested-type-name-match constraints by first keeping them out of
the connected-components algorithm used to compute same-type
constraints within an equivalence class. Then, detect self-derived
nested-type-name-match constraints by determining whether any of the
ancestor potential archetypes are in the same equivalence class... and
redundant with the edge that makes the ancestor potential archetypes
equivalent. Remove such self-derived edges, and treat all other
nested-type-name-match edges as derived, providing a minimized
signature.

Fixes SR-5841, SR-5601, and SR-5610

DougGregor committed:
[Sequence] Make Sequence.SubSequence conform to Sequence.

Addressed ABI FIXME’s #4, #5, #104 and #105, making Sequence’s
SubSequence conform to Sequence, with the same element type, and for
which the SubSequence of a SubSequence is the same SubSequence.

Fixes SR-318 / rdar://problem/31418206.

DougGregor committed:
[GSB] Split out the operation to expand the requirement signature of a conformance.

NFC for now; we are going to re-use this code.

moiseev commented on stdlib/public/core/SequenceWrapper.swift:
```diff
@@ -110,10 +106,6 @@ extension _SequenceWrapper
     return try _base.prefix(while: predicate)
   }
   
-  public func suffix(_ maxLength: Int) -> SubSequence {
```
Is it no longer necessary?

DougGregor replied:
It's redundant, because I merged two extensions into a single extension. The extension of `_SequenceWrapper`

    where SubSequence == Base.SubSequence, Element == Base.Element

is just an extension of `_SequenceWrapper` now.

DougGregor committed:
[GSB] Process same-type requirements needed to collapse same-type components.

Same-type constraints are rederived based on the potential archetypes
within an equivalence class and the same-type constraints that link
them. In some cases, there may be parts that connect those same-type
constraints that are stored in "delayed" requirements (for which one
or both end-points have not been realized in a potential archetype);
consider those as well for connectedness.

In a sense, this is an artifact of the generic signature builder's
approach of realizing potential archetypes too readily, and this can
likely be simplified in the future. For now, it's important for the
minimization of same-type constraints in generic signatures.

DougGregor committed:
[GSB] Dump delayed requirements to aid in debugging. NFC

DougGregor committed:
[GSB] Collapse same-type components using delayed requirements.

Perform the actual "collapse" of same-type components by looking
through the delayed requirements of an equivalence class.

DougGregor committed:
[GSB] Collapse structurally-equivalent potential archetypes.

When there are to "structurally" equivalent potential archetypes in
different components of an equivalence class, as determined by their
nested-name structure, collapse the two components. This addresses the
remaining known issues due to the eager explosion of potential
archetypes in the generic signature builder.

DougGregor committed:
[Stdlib] Eliminate the _UnicodeEncoding_ protocol.

This protocol was only used to fake recursive protocol conformances;
collapse it into _UnicodeEncoding.

DougGregor committed:
“Sanitize” requirements before checking inferred type witnesses.

When we check type inferred type witnesses against the requirement
signature of the protocol, first "sanitize" the requirements to remap
any dependent member types into the equivalent dependent member types
within that protocol, dropping any requirements that can't be
remapped. This eliminates some extraneous recursion introduced by
checking requirements, and isolates us somewhat from the specific
choice of "canonical" associated types.

DougGregor commented:
@moiseev , @airspeedswift , and @dabrahams might be interested in this. Only the last commit is truly part of this pull request; the rest is the compiler work in https://github.com/apple/swift/pull/11751.

DougGregor commented:
@swift-ci please test

DougGregor commented:
I'm reverting this in https://github.com/apple/swift/pull/11921 and will collect all of the recursive constraints for the standard library into a separate PR, to land when it doesn't affect compile-time performance too horribly.
